pytorch_lightning/overrides/data_parallel.py

```diff
@@ -45,133 +40,103 @@ def _find_tensors(obj):  # pragma: no-cover
 
-        # remove all the meta info
-        meta = outputs[0]['meta']
-        for i, output in enumerate(outputs):
-            del output['meta']
+class LightningDistributedDataParallel(DistributedDataParallel):
 
-        outputs = self.gather(outputs)
+    def __init__(self, module: LightningModule, *args, **kwargs):
+        warnings.warn(
+            "The usage of `LightningDistributedDataParallel` is deprecated since v1.2 and will be removed in v1.4."
+            " From now on we recommend to directly subclass `torch.nn.parallel.DistributedDataParallel`.",
+            DeprecationWarning
+        )
+        super().__init__(LightningDistributedModule(module), *args, **kwargs)
 
-        result = original_class()
 
-        result.update(outputs)
-        result['meta'] = meta
-        return result
+class LightningModuleWrapperBase(torch.nn.Module):
 
-    def gather(self, outputs):
-        r"""
-        Override the gather method to support python scalars as well.
+    def __init__(self, pl_module: LightningModule):
         """
-        def gather_map(outputs):
-            elem = outputs[0]
-            elem_type = type(elem)
-
-            if isinstance(elem, torch.Tensor):
-                return Gather.apply(self.output_device, self.dim, *outputs)
+        Wraps the user's LightningModule and redirects the forward call to the appropriate
+        method, either ``training_step``, ``validation_step`` or ``test_step``.
+        Inheriting classes may also modify the inputs or outputs of forward.
 
-            if elem is None:
-                return None
+        Args:
+            pl_module: the model to wrap
+        """
+        super().__init__()
+        self.module = pl_module
 
-            if isinstance(elem, Mapping):
-                if not all((len(elem) == len(d) for d in outputs)):
-                    raise ValueError('All dicts must have the same number of keys')
-                return elem_type(((k, gather_map([d[k] for d in outputs]))
-                                  for k in elem))
+    def forward(self, *inputs, **kwargs):
+        running_stage = self.module.running_stage
 
-            if isinstance(elem, Iterable) and not isinstance(elem, str):
-                return elem_type(map(gather_map, zip(*outputs)))
+        if running_stage == RunningStage.TRAINING:
+            output = self.module.training_step(*inputs, **kwargs)
+            warn_if_output_is_none(output, "training_step")
+        elif running_stage == RunningStage.TESTING:
+            output = self.module.test_step(*inputs, **kwargs)
+            warn_if_output_is_none(output, "test_step")
+        elif running_stage == RunningStage.EVALUATING:
+            output = self.module.validation_step(*inputs, **kwargs)
+            warn_if_output_is_none(output, "validation_step")
+        else:
+            output = self.module.predict(*inputs, **kwargs)
 
-            return outputs
+        return output
```

Comment:
n00b question: does data parallel work outside of training/testing?

Reply:
yes, outside of training and testing it behaves like the regular forward, i.e., it will pass the arguments to LightningModule.predict which defaults to redirecting to LightningModule.forward, unless user overrides that.